### PR TITLE
Node partitioned mesh

### PR DESCRIPTION
**Pull request by endJunction**
This is a split-off from ufz/ogs#580

Both new functions are indeed used in the above mentioned PR.

Unfortunatly jenkins went down some time ago and I can only hope for correct compilation on other platforms but linux.
@wenqing Please confirm this code is working for you, then someone merge it.

### CODE REVIEW

File: MeshLib/NodePartitionedMesh.h [hunk 1]
```diff
@@ -115,6 +120,19 @@ class NodePartitionedMesh : public Mesh
...
+        std::size_t getMaximumNConnectedNodesToNode() const
+        {
+            std::vector<Node *>::const_iterator it_max_ncn = std::max_element(
+                _nodes.cbegin(), _nodes.cend(),
+                [](Node const *const node_a, Node const *const node_b)
+                {
+                    return (node_a->getConnectedNodes().size() <
+                            node_b->getConnectedNodes().size());
+                });
+            return (*it_max_ncn)->getConnectedNodes().size() + 1;
```
Comment (TomFischer) [inline MeshLib/NodePartitionedMesh.h]: Why isn't the `size()` as return value enough, i.e. why is it necessary to add one here?

Comment (wenqing) [inline MeshLib/NodePartitionedMesh.h]: The number of connected nodes should include the node itself. Besides, this number is used to preallocate memory for PETSc matrix, the number plus one guarantees an enough space for the sparsity of the matrix arising from FEM.

Comment (endJunction) [inline MeshLib/NodePartitionedMesh.h]: Thanks Wenqing; I added a comment for this.

**Comment by wenqing** [comment]
All tests run except a compilation error caused by
```
#ifdef USE_PETSC
	for (auto p_it = project.processesBegin(); p_it != project.processesEnd(); ++p_it)
	{
		(*p_it)->releaseEquationMemory();
	}
...
#endif
```
which should be in #580, but has been committed to master via #831. Anyway, there is not any test for ogs#PETSc in the master, and everything looks good. Therefore, this PR can be merged.

File: MeshLib/NodePartitionedMesh.h [hunk 2]
```diff
@@ -115,6 +120,20 @@ class NodePartitionedMesh : public Mesh
...
+        std::size_t getMaximumNConnectedNodesToNode() const
+        {
+            std::vector<Node *>::const_iterator it_max_ncn = std::max_element(
+                _nodes.cbegin(), _nodes.cend(),
+                [](Node const *const node_a, Node const *const node_b)
+                {
+                    return (node_a->getConnectedNodes().size() <
+                            node_b->getConnectedNodes().size());
+                });
+            // Return the number of connected nodes +1 for the node itself.
```
Comment (wenqing) [inline MeshLib/NodePartitionedMesh.h]: // Return the number of the connected nodes and the node itself.

**Comment by bilke** [comment]
:+1:

**Pull request merged**